File: util/java/src/com/db/util/KeyValuePair.java
/*
 * Copyright (c) 2005-2006 Digital Bazaar, Inc.  All rights reserved.
 */
package com.db.util;

import java.util.Comparator;

/**
 * A simple class for storing key-value pairs.
 * 
 * @author Dave Longley
 */
public class KeyValuePair implements Comparator
{
   /**
    * The key.
    */
   protected Object mKey;
   
   /**
    * The value.
    */
   protected Object mValue;
   
   /**
    * Creates a new KeyValuepair with the given key and value.
    * 
    * @param key the key.
    * @param value the value.
    */
   public KeyValuePair(Object key, Object value)
   {
      mKey = key;
      mValue = value;
   }
   
   /**
    * Returns the key.
    *
    * @return the key.
    */
   public Object getKey()
   {
      return mKey;
   }
   
   /**
    * Returns the value.
    * 
    * @return the value.
    */
   public Object getValue()
   {
      return mValue;
   }
   
   /**
    * Compares two key-value pairs. Compares on the value of the key-value
    * pair. If the first argument has a lesser value than the second,
    * a negative integer is returned. If the first argument is equal to
    * the second, 0 is returned. If the first argument is greater than
    * the second, than a positive integer is returned.
    * 
    * @param o1 the first object to be compared.
    * @param o2 the second object to be compared.
    * @return a negative integer, zero, or a positive integer as the
    *         first argument is less than, equal to, or greater than the
    *         second. 
    * @throws ClassCastException if the arguments' types prevent them from
    *          being compared by this Comparator.
    */
   public int compare(Object o1, Object o2)
   {
      int rval = 0;
      
      KeyValuePair kvp1 = (KeyValuePair)o1;
      KeyValuePair kvp2 = (KeyValuePair)o2;
      
      String value1 = kvp1.getValue().toString();
      String value2 = kvp2.getValue().toString();
      
      rval = value1.compareTo(value2);
      
      if(rval != 0)
      {
         rval = (rval < 0) ? -1 : 1;
      }
      
      return rval;
   }   
   
   /**
    * A key-value pair is equal to another key-value pair if
    * its key and value are equal.
    * 
    * @param obj another key-value pair to compare to.
    */
   public boolean equals(Object obj)
   {
      boolean rval = false;
      
      if(obj != this)
      {
         /*if(obj != null)
         {
            if(obj instanceof KeyValuePair)
            {
               KeyValuePair kvp = (KeyValuePair)obj;
               if(kvp.getKey().equals(getKey()) &&
                  kvp.getValue().equals(getValue()))
               {
                  rval = true;
               }
            }
         }*/
      }
      else
      {
         rval = true;
      }
      
      return rval;
   }
}

File: obsolete/common/csharp/DB.Common/KeyValuePair.cs
/*
 * Copyright (c) 2005 Digital Bazaar, Inc.  All rights reserved.
 */
using System;
using System.Collections;
using System.IO;
using System.Text;
using System.Xml;

using DB.Common.Logging;

namespace DB.Common
{
   /// <summary>
   /// A simple class for storing key-value pairs.
   /// </summary>
   ///
   /// <author>Dave Longley</author>
   public class KeyValuePair : IComparer
   {
      /// <summary>
      /// The key.
      /// </summary>
      protected object mKey;
      
      /// <summary>
      /// The value.
      protected object mValue;
      
      /// <summary>
      /// Creates a new KeyValuePair with the given key and value.
      /// </summary>
      ///
      /// <param name="key">the key.</param>
      /// <param name="val">the value.</param>
      public KeyValuePair(object key, object val)
      {
         if(key != null)
         {
            mKey = key;
         }
         else
         {
            mKey = "null";
         }
         
         if(val != null)
         {
            mValue = val;
         }
         else
         {
            mValue = "null";
         }
      }
      
      /// <summary>
      /// Gets the key of this key-value pair.
      /// </summary>
      public virtual object Key
      {
         get
         {
            return mKey;
         }
      }
      
      /// <summary>
      /// Gets the value of this key-value pair.
      /// </summary>
      public virtual object Value
      {
         get
         {
            return mValue;
         }
      }
      
      /// <summary>
      /// Compares two key-value pairs. Compares on the value of the key-value
      /// pair. If the first argument has a lesser value than the second,
      /// a negative integer is returned. If the first argument is equal to
      /// the second, 0 is returned. If the first argument is greater than
      /// the second, than a positive integer is returned.
      /// </summary>
      /// 
      /// <param name="o1">the first object to be compared.</param>
      /// <param name="o2">the second object to be compared.</param>
      /// <returns>a negative integer, zero, or a positive integer as the
      /// first argument is less than, equal to, or greater than the
      /// second.</returns> 
      public virtual int Compare(object o1, object o2)
      {
         int rval = 0;
         
         KeyValuePair kvp1 = (KeyValuePair)o1;
         KeyValuePair kvp2 = (KeyValuePair)o2;
         
         if(kvp1.Value != null && kvp1.Value.Equals(kvp2.Value))
         {
            rval = 0;
         }
         else if(kvp1.Value != null && kvp2.Value != null)
         {
            string value1 = kvp1.Value.ToString();
            string value2 = kvp2.Value.ToString();
         
            rval = value1.CompareTo(value2);
         
            if(rval != 0)
            {
               rval = (rval < 0) ? -1 : 1;
            }
         }
         
         return rval;
      }
      
      /// <summary>
      /// A key-value pair is equal to another key-value pair if
      /// its key and value are equal.
      /// </summary>
      /// 
      /// <param name="kvp">another key-value pair to compare to.</param>
      public virtual bool Equals(KeyValuePair kvp)
      {
         bool rval = false;
         
         if(kvp != this)
         {
            /*if(kvp != null)
            {
               if(kvp.Key.Equals(Key) &&
                  kvp.Value.Equals(Value))
               {
                  rval = true;
               }
            }*/
         }
         else
         {
            rval = true;
         }
         
         return rval;
      }
   }
}
